### PR TITLE
PWGHF: new task for W/Z->e in mid-rapidity

### PR DESCRIPTION
**Pull request by sashingo**
PWGHF: a new task for W/Z decay to electron in midrapidity in Run3

### CODE REVIEW

File: PWGHF/HFL/Tasks/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,11 @@
 # granted to it by virtue of its status as an Intergovernmental Organization
 # or submit itself to any jurisdiction.
 
+o2physics_add_dpl_workflow(task-electron-weakboson
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/CMakeLists.txt:12]:
```suggestion
o2physics_add_dpl_workflow(task-electron-weak-boson
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+///
+/// \ task for WeakBoson (W/Z) based on electron in mid-rapidity
+///        
+/// \S. Sakai & S. Ito
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:14]: Fix the documentation

File: PWGHF/HFL/Tasks/CMakeLists.txt [hunk 3]
```diff
@@ -33,3 +38,4 @@ o2physics_add_dpl_workflow(task-single-muon-source
                     SOURCES taskSingleMuonSource.cxx
                     PUBLIC_LINK_LIBRARIES O2Physics::AnalysisCore
                     COMPONENT_NAME Analysis)
+
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/CMakeLists.txt:41]:
```suggestion
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+#include "Common/DataModel/TrackSelectionTables.h"
+#include "Common/DataModel/PIDResponse.h"
+#include "Framework/ASoAHelpers.h"
+
+#include "EMCALBase/Geometry.h"
+#include "EMCALCalib/BadChannelMap.h"
+#include "PWGJE/DataModel/EMCALClusters.h"
+#include "DataFormatsEMCAL/Cell.h"
+#include "DataFormatsEMCAL/Constants.h"
+#include "DataFormatsEMCAL/AnalysisCluster.h"
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:28]: Fix the order

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+using selectedClusters = o2::soa::Filtered<o2::aod::EMCALClusters>;
+
+// PbPb
+using TrackEle = o2::soa::Filtered<o2::soa::Join<o2::aod::Tracks, o2::aod::FullTracks, o2::aod::TracksExtra_001, o2::aod::TracksDCA, o2::aod::TrackSelection, o2::aod::pidTPCFullEl>>;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:37]: There is no reason to declare these outside the `struct`.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+#include "PWGJE/DataModel/EMCALClusters.h"
+#include "DataFormatsEMCAL/Cell.h"
+#include "DataFormatsEMCAL/Constants.h"
+#include "DataFormatsEMCAL/AnalysisCluster.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+using selectedClusters = o2::soa::Filtered<o2::aod::EMCALClusters>;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:34]: Respect naming conventions.
```suggestion
using SelectedClusters = o2::soa::Filtered<o2::aod::EMCALClusters>;
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  Preslice<o2::aod::EMCALAmbiguousClusterCells> perClusterAmb = o2::aod::emcalclustercell::emcalambiguousclusterId;
+  PresliceUnsorted<o2::aod::EMCALMatchedTracks> perClusterMatchedTracks = o2::aod::emcalmatchedtrack::trackId;
+
+  Configurable<int> nBinsPt{"nBinsPt", 100, "N bins in pt histo"};
+  Configurable<int> nBinsE{"nBinsE", 100, "N bins in E histo"};
+
+  // event filter
+  Filter eventFilter = (o2::aod::evsel::sel8 == true);
+  Configurable<float> vtxZ{"vtxZ", 10.f, ""};
+  Filter posZFilter = (nabs(o2::aod::collision::posZ) < vtxZ);
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:54]: Respect order of members.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+  // track cuts
+  Filter filter_globalTr = requireGlobalTrackInFilter();  
+
+  Configurable<float> etalow{"etalow", -0.6f, ""};
+  Configurable<float> etaup{"etaup", 0.6f, ""};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:60]: Pay attention to proper names.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+  Configurable<float> tpcchi2_cut{"tpcchi2_cut", 4.0f, "tpc chi2 cut"};
+  Configurable<float> itsNcl_cut{"itsNcl_cut", 2.0f, "its # of cluster cut"};
+  Configurable<float> tpcNcl_cut{"tpcNcl_cut", 100.0f, "tpc # if cluster cut"};
+  Configurable<float> tpcNclCr_cut{"tpcNclCr_cut", 100.0f, "tpc # of crossedRows cut"};
+
+  // cluster cut
+  Configurable<int> mClusterDefinition{"clusterDefinition", 10, "cluster definition to be selected, e.g. 10=kV3Default"};
+  Configurable<float> minTime{"minTime", -25., "Minimum cluster time for time cut"}; 
+  Configurable<float> maxTime{"maxTime", +20., "Maximum cluster time for time cut"}; 
+  Configurable<float> minM02{"minM02", 0.1, "Minimum M02 for M02 cut"};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:77]: Fix the names

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    histos.add("TrMatchHistogram_mim", "TrMatchHistogram_mim", kTH2F, {{axisdPhi},{axisdEta}});
+    histos.add("MatchPhiHistogram", "MatchPhiHistogram", kTH2F, {{axisPhi},{axisPhi}});
+    histos.add("MatchEtaHistogram", "MatchEtaHistogram", kTH2F, {{axisEta},{axisEta}});
+    histos.add("EopHistogram", "EopHistogram", kTH2F, {{axisPt},{axisEop}});
+    histos.add("EopNsigTPCHistogram", "EopNsigTPCHistogram", kTH2F, {{axisNsigma},{axisEop}});
+    histos.add("EMCtimeHistogram", "EMCtimeHistogram", kTH1F, {axisEMCtime});
+  }
+
+  //void process(soa::Filtered<aod::Collisions>::iterator const& collision, selectedClusters const& clusters, TrackEle const& tracks, o2::aod::EMCALMatchedTracks const& matchedtracks)
+  void process(soa::Filtered<aod::Collisions>::iterator const& collision, selectedClusters const& emcClusters, TrackEle const& tracks, o2::aod::EMCALMatchedTracks const& matchedtracks)
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:132]: One argument per line

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    histos.fill(HIST("hEventCounter"),0.5);
+
+    //LOGF(info, "Collision index : %d", collision.index());
+    //LOGF(info, "Number of tracks: %d", tracks.size());
+    //LOGF(info, "Number of clusters: %d", clusters.size());
+    //LOGF(info, "Number of clusters: %d", emcClusters.size());
+
+    histos.fill(HIST("ZvtxHistogram"), collision.posZ());
+
+    for (auto& track : tracks) {
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:143]: Missing `const`.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+      histos.fill(HIST("etaHistogram"), track.eta());
+      histos.fill(HIST("ITS_Chi2_Hist"), track.itsChi2NCl());
+      histos.fill(HIST("TPC_Chi2_Hist"), track.tpcChi2NCl());
+      histos.fill(HIST("TPC_NCls_Hist"), track.tpcNClsFound());
+      histos.fill(HIST("ITS_NCls_Hist"), track.itsNCls());
+      histos.fill(HIST("TPC_NClsCrossedRows_Hist"), track.tpcNClsCrossedRows());
+
+      if(TMath::Abs(track.eta())>0.6)continue;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:152]: No magic numbers

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+      if(track.tpcNClsCrossedRows()<tpcNclCr_cut)continue;
+      if(TMath::Abs(track.dcaXY())>dcaxy_cut)continue;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:154]:
```suggestion
      if(std::abs(track.eta())>0.6)continue;
      if(track.tpcNClsCrossedRows()<tpcNclCr_cut)continue;
      if(std::abs(track.dcaXY())>dcaxy_cut)continue;
```

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by sashingo** [comment]
Dear Vit,
thank you very much for reviewing the commit.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 8 above)
Comment (sashingo) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:60]: keep as it is

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 9 above)
Comment (sashingo) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx]: keep as it is

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 7 above)
Comment (sashingo) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:56]: I don't understand the comment

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 14]
```diff
@@ -8,38 +8,40 @@
 // In applying this license CERN does not waive the privileges and immunities
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
-///
-/// \ task for WeakBoson (W/Z) based on electron in mid-rapidity
-///        
-/// \S. Sakai & S. Ito
+/// \file taskElectronWeakNoson.cxx
+/// \briff task for WeakBoson (W/Z) based on electron in mid-rapidity
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:12]:
```suggestion
/// \brief task for WeakBoson (W/Z) based on electron in mid-rapidity
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 7 above)
Comment (vkucera) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:56]: Please don't mark comments as resolved when you did not address them. For consistency, group the member by type and follow the conventions for their order.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 8 above)
Comment (vkucera) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:60]: Please don't mark comments as resolved when you did not address them. The current names are ambiguous (eta of what? what is the meaning of up/low?), they don't follow naming conventions and the variables are not documented. Please fix it.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 9 above)
Comment (vkucera) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx]: not addressed

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 10 above)
Comment (vkucera) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx]: not addressed

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx (hunk 4 above)
Comment (vkucera) [reply PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx]: not addressed
Include O2Physics headers after O2.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+/// \file taskElectronWeakNoson.cxx
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:11]:
```suggestion
// or submit itself to any jurisdiction.

/// \file taskElectronWeakNoson.cxx
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+  Configurable<float> minM02{"minM02", 0.1, "Minimum M02 for M02 cut"};
+  Configurable<float> maxM02{"maxM02", 0.9, "Maximum M02 for M02 cut"};
+  Configurable<float> minM20{"minM20", 0.1, "Minimum M20 for M20 cut"};
+  Configurable<float> maxM20{"maxM20", 0.6, "Maximum M20 for M20 cut"};
+  Configurable<float> MatchR_cut{"MatchR_cut", 0.1, "cluster - track matching cut"};
+
+  Filter clusterDefinitionSelection = (o2::aod::emcalcluster::definition == mClusterDefinition) && (o2::aod::emcalcluster::time >= minTime) && (o2::aod::emcalcluster::time <= maxTime) && (o2::aod::emcalcluster::m02 > minM02) && (o2::aod::emcalcluster::m02 < maxM02);
+
+  // Histogram registry: an object to hold your histograms
+  HistogramRegistry histos{"histos", {}, OutputObjHandlingPolicy::AnalysisObject};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:86]:
```suggestion
  // Histogram registry: an object to hold your histograms
  HistogramRegistry registry{"registry"};
```
Keep it simple.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+    histos.add("EnergyHistogram", "EnergyHistogram", kTH1F, {axisE});
+    histos.add("M02Histogram", "M02Histogram", kTH2F, {{axisNsigma}, {axisM02}});
+    histos.add("M20Histogram", "M20Histogram", kTH2F, {{axisNsigma}, {axisM02}});
+    histos.add("TrMatchHistogram", "TrMatchHistogram", kTH2F, {{axisdPhi}, {axisdEta}});
+    histos.add("TrMatchHistogram_mim", "TrMatchHistogram_mim", kTH2F, {{axisdPhi}, {axisdEta}});
+    histos.add("MatchPhiHistogram", "MatchPhiHistogram", kTH2F, {{axisPhi}, {axisPhi}});
+    histos.add("MatchEtaHistogram", "MatchEtaHistogram", kTH2F, {{axisEta}, {axisEta}});
+    histos.add("EopHistogram", "EopHistogram", kTH2F, {{axisPt}, {axisEop}});
+    histos.add("EopNsigTPCHistogram", "EopNsigTPCHistogram", kTH2F, {{axisNsigma}, {axisEop}});
+    histos.add("EMCtimeHistogram", "EMCtimeHistogram", kTH1F, {axisEMCtime});
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:128]: The "Histogram" strings are not adding any information.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+      if (track.pt() < mimpT_cut)
+        continue;
+      histos.fill(HIST("ptHistogram"), track.pt());
+      histos.fill(HIST("TPCElHistogram"), track.p(), track.tpcNSigmaEl());
+
+      // track - match
+
+      if (emcClusters.size() < 1)
+        continue;
+      if (track.phi() < 1.39 || track.phi() > 3.15)
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:175]: No magic numbers.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+  // cluster cut
+  Configurable<int> mClusterDefinition{"mClusterDefinition", 10, "cluster definition to be selected, e.g. 10=kV3Default"};
+  Configurable<float> minTime{"minTime", -25., "Minimum cluster time for time cut"};
+  Configurable<float> maxTime{"maxTime", +20., "Maximum cluster time for time cut"};
+  Configurable<float> minM02{"minM02", 0.1, "Minimum M02 for M02 cut"};
+  Configurable<float> maxM02{"maxM02", 0.9, "Maximum M02 for M02 cut"};
+  Configurable<float> minM20{"minM20", 0.1, "Minimum M20 for M20 cut"};
+  Configurable<float> maxM20{"maxM20", 0.6, "Maximum M20 for M20 cut"};
+  Configurable<float> MatchR_cut{"MatchR_cut", 0.1, "cluster - track matching cut"};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:81]:
```suggestion
  Configurable<float> minTime{"minTime", -25., "Minimum cluster time"};
  Configurable<float> maxTime{"maxTime", +20., "Maximum cluster time"};
  Configurable<float> minM02{"minM02", 0.1, "Minimum M02"};
  Configurable<float> maxM02{"maxM02", 0.9, "Maximum M02"};
  Configurable<float> minM20{"minM20", 0.1, "Minimum M20"};
  Configurable<float> maxM20{"maxM20", 0.6, "Maximum M20"};
  Configurable<float> MatchR_cut{"MatchR_cut", 0.1, "cluster - track matching cut"};
```
Do not repeat yourself.
The name of the cut variable and its documentation should be clear enough to tell how the cut is applied. That is not the case for `MatchR_cut`

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+        int nmatch = 0;
+        for (const auto& match : tracksofcluster) {
+          if (match.emcalcluster_as<SelectedClusters>().time() < minTime || match.emcalcluster_as<SelectedClusters>().time() > maxTime)
+            continue;
+          if (match.emcalcluster_as<SelectedClusters>().m02() < minM02 || match.emcalcluster_as<SelectedClusters>().m02() > maxM02)
+            continue;
+
+          double emc_m20 = match.emcalcluster_as<SelectedClusters>().m20();
+          double emc_m02 = match.emcalcluster_as<SelectedClusters>().m02();
+          double emc_energy = match.emcalcluster_as<SelectedClusters>().energy();
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:195]: Do you need `double`?

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+          // LOG(info) << "tr phi1 = " << track.phi();
+          // LOG(info) << "emc phi = " << emc_phi;
+          if (nmatch == 0) {
+            double dPhi = match.track_as<TrackEle>().phi() - emc_phi;
+            double dEta = match.track_as<TrackEle>().eta() - emc_eta;
+
+            histos.fill(HIST("MatchPhiHistogram"), emc_phi, match.track_as<TrackEle>().phi());
+            histos.fill(HIST("MatchEtaHistogram"), emc_eta, match.track_as<TrackEle>().eta());
+
+            double R = sqrt(pow(dPhi, 2) + pow(dEta, 2));
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:209]:
```suggestion
            double R = std::sqrt(std::pow(dPhi, 2) + std::pow(dEta, 2));
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+          double emc_m02 = match.emcalcluster_as<SelectedClusters>().m02();
+          double emc_energy = match.emcalcluster_as<SelectedClusters>().energy();
+          double emc_phi = match.emcalcluster_as<SelectedClusters>().phi();
+          double emc_eta = match.emcalcluster_as<SelectedClusters>().eta();
+          double emc_time = match.emcalcluster_as<SelectedClusters>().time();
+          // LOG(info) << "tr phi0 = " << match.track_as<TrackEle>().phi();
+          // LOG(info) << "tr phi1 = " << track.phi();
+          // LOG(info) << "emc phi = " << emc_phi;
+          if (nmatch == 0) {
+            double dPhi = match.track_as<TrackEle>().phi() - emc_phi;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:203]: You should constrain `dPhi` to be between -π and +π for distance calculations. See `RecoDecay::constrainAngle`.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+            if (match.track_as<TrackEle>().tpcNSigmaEl() > -1.0 && match.track_as<TrackEle>().tpcNSigmaEl() < 3) {
+              histos.fill(HIST("EopHistogram"), match.track_as<TrackEle>().pt(), eop);
+            }
+          }
+
+          nmatch++;
+        }
+      }
+
+      if (Rmim < 10.0) {
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:235]: No magic numbers

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+            if (R < MatchR_cut)
+              continue;
+
+            double eop = emc_energy / match.track_as<TrackEle>().p();
+            // LOG(info) << "E/p" << eop;
+            histos.fill(HIST("EopNsigTPCHistogram"), match.track_as<TrackEle>().tpcNSigmaEl(), eop);
+            histos.fill(HIST("M02Histogram"), match.track_as<TrackEle>().tpcNSigmaEl(), emc_m02);
+            histos.fill(HIST("M20Histogram"), match.track_as<TrackEle>().tpcNSigmaEl(), emc_m20);
+            if (match.track_as<TrackEle>().tpcNSigmaEl() > -1.0 && match.track_as<TrackEle>().tpcNSigmaEl() < 3) {
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:226]: No magic numbers

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by sashingo** [comment]
recommitted after implementing suggestions by Vit

**Comment by vkucera** [comment]
Thanks @sashingo for implementing the comments.
Please try to stick to the [O2 naming conventions](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=General_Naming_Rules#General_Naming_Rules) and use the lowerCamelCase style for the names of variables and functions. It improves readability of your code and helps others understand it.

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+struct HfTaskElectronWeakBoson {
+
+  using SelectedClusters = o2::soa::Filtered<o2::aod::EMCALClusters>;
+
+  // PbPb
+  using TrackEle = o2::soa::Filtered<o2::soa::Join<o2::aod::Tracks, o2::aod::FullTracks, o2::aod::TracksExtra_001, o2::aod::TracksDCA, o2::aod::TrackSelection, o2::aod::pidTPCFullEl>>;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:42]:
```suggestion
  using TrackEle = o2::soa::Filtered<o2::soa::Join<o2::aod::Tracks, o2::aod::FullTracks, o2::aod::TracksExtra_001, o2::aod::TracksDCA, o2::aod::TrackSelection, o2::aod::pidTPCFullEl>>;
```
In general, it's a bad idea to subscribe to a specific version of a table. What is your reason for doing so?

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+  Configurable<float> tpcchi2_cut{"tpcchi2_cut", 4.0f, "tpc chi2 cut"};
+  Configurable<float> itsNcl_cut{"itsNcl_cut", 2.0f, "its # of cluster cut"};
+  Configurable<float> tpcNcl_cut{"tpcNcl_cut", 100.0f, "tpc # if cluster cut"};
+  Configurable<float> tpcNclCr_cut{"tpcNclCr_cut", 100.0f, "tpc # of crossedRows cut"};
+  Configurable<float> tpcNsiglow_cut{"tpcNsiglow_cut", -1.0, "tpc Nsig lower cut"};
+  Configurable<float> tpcNsigup_cut{"tpcNsigup_cut", 3.0, "tpc Nsig upper cut"};
+
+  Configurable<float> emcacc_phimin{"emcacc_phimin", 1.39, "Maximum M20"};
+  Configurable<float> emcacc_phimax{"emcacc_phimax", 3.36, "Maximum M20"};
+  Configurable<int> mClusterDefinition{"mClusterDefinition", 10, "cluster definition to be selected, e.g. 10=kV3Default"};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:69]: What does `m` in `mClusterDefinition` mean?

**Comment by sashingo** [comment]
Hello, I think I address all the comments by Vit.
Please check and approve if it's ok.

**Review by NicoleBastid — APPROVED** [review]

**Review by NicoleBastid — DISMISSED** [review]

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 27]
```diff
@@ -0,0 +1,259 @@
...
+      auto tracksofcluster = matchedtracks.sliceBy(perClusterMatchedTracks, track.globalIndex());
+
+      // LOGF(info, "Number of matched track: %d", tracksofcluster.size());
+
+      double Rmim = 999.9;
+      double dPhi_mim = 999.9;
+      double dEta_mim = 999.9;
+
+      if (tracksofcluster.size()) {
+        int nmatch = 0;
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:191]:
```suggestion
        int nMatch = 0;
```

File: PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx [hunk 28]
```diff
@@ -0,0 +1,259 @@
...
+      }
+
+    } // end of track loop
+  }
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  return WorkflowSpec{
+    adaptAnalysisTask<HfTaskElectronWeakBoson>(cfgc)};
```
Comment (vkucera) [inline PWGHF/HFL/Tasks/taskElectronWeakBoson.cxx:258]:
```suggestion
  return WorkflowSpec{adaptAnalysisTask<HfTaskElectronWeakBoson>(cfgc)};
```

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @sashingo , thanks for implementing the comments. Please seem some further suggestions.

**Comment by sashingo** [comment]
Hi Vit,
thank you for further suggestions.
I implemented them

**Review by vkucera — APPROVED** [review]
Thanks @sashingo for implementing all my comments.

**Pull request merged**